src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs

```diff
@@ -1358,7 +1358,123 @@ private struct MessageHeader
                 public bool Fin;
                 public long PayloadLength;
             }
+
+            /// <summary>
+            /// A custom network stream that stores and reuses a single SocketAsyncEventArgs instance
+            /// for reads and a single SocketAsyncEventArgs instance for writes.  This limits it to
+            /// supporting a single read and a single write at a time, but with much less per-operation
+            /// overhead than with System.Net.Sockets.NetworkStream.
+            /// </summary>
+            private sealed class AsyncEventArgsNetworkStream : NetworkStream
+            {
+                private readonly Socket _socket;
+                private readonly SocketAsyncEventArgs _readArgs;
+                private readonly SocketAsyncEventArgs _writeArgs;
+
+                private AsyncTaskMethodBuilder<int> _readTcs;
+                private AsyncTaskMethodBuilder _writeTcs;
```

Comment:
Originally these were TaskCompletionSources.  I changed them to be AsyncTaskMethodBuilders for perf reasons.  ATMB is a struct whereas TCS is a class, but otherwise logically they're very similar: both create a task and let you manually complete it.  ATMB is trickier to use, but in this case its use seemed warranted as it allowed me to avoid an extra allocation per operation.  In any event, I've renamed them.